### ISSUE TITLE
jwasmr (or OpenWatcom) broken since at least 2.18

### ISSUE THREAD

**Issue by boeckmann** [SvarDOS/bugz#94]
jwasmr crashes when launched with SvarCOM under EDR and FreeDOS kernel. I do not think it is actually SvarCOM related, but running under SvarCOM seems to increase the chance this is happening. Problem occurs with the "official" JWasmr 2.18 binary and with my self-built master versions (OpenWatcom v2).

I have a hard time debugging this, because it runs when compiled with debug info. I tried to debug the release version. This crashes in lDebug. But if I try to isolate the error by utilizing lDebug trace mode, it runs...

Screenshots and disk image attached...

<img width="832" alt="Bildschirmfoto 2024-08-10 um 12 28 19" src="https://github.com/user-attachments/assets/d2e6bb90-cf94-4926-831a-62ecf84c3b6d">

<img width="832" alt="Bildschirmfoto 2024-08-10 um 12 28 49" src="https://github.com/user-attachments/assets/2fdf7af4-1578-4c9a-a7cf-5f35098bb0da">
[jwasmr.img.zip](https://github.com/user-attachments/files/16569246/jwasmr.img.zip)

**Comment by boeckmann** [comment]
[jwasmr.img.zip](https://github.com/user-attachments/files/16569278/jwasmr.img.zip)

**Comment by boeckmann** [comment]
I managed to produce a semingly working binary. The original makefile builds a large library and links it against the main file. I changed this to directly link the binary. This results in a changed segment order, putting the _TEXT segment containing the default library code as the last code segment. I do not know if the watcom toolchain / library actually depends on this in the large memory model. But it obviously has some effect. Maybe simply some other memory corruption takes place now. But until now I found no obvious flaw in the JWasmr code that would otherwise explain the problem, as the code crashed more or less directly after a few function calls into main()...

**Comment by boeckmann** [comment]
As the JWasm data is more than 64k, a few OpenWatcom C options are used in an effort to reduce the data segment size of the 16 bit binary, specifically -zc to put constants into the code segments. Guess: together with the somewhat unorthodox segment ordering of the original binary, this MAY have lead to some edge cases resulting in a broken binary.

**Comment by boeckmann** [comment]
I tested two build configurations for the real mode JWasm.

1. without the `-zc` (consts NOT put into code segments), but a `-zt=1024` (putting big structures into FARDATA to keep _DATA smaller 64K) and removing the .LIB step, directly linking to an executable.
2. without `-zc`, but with the interim JWasm .LIB creation.

Both work at least to the point, that under EDR I can create a working EDR kernel. So it seems to me that -zc and _TEXT not being the last code segment indeed seems to be troublesome. Not sure what to make about this.

**Comment by boeckmann** [comment]
I filed a bug report at https://github.com/Baron-von-Riedesel/JWasm/issues/29

**Comment by boeckmann** [comment]
Baron von Riedesel thinks this is a QEMU bug, as this runs with KVM enabled QEMU and / or slightly altered JWasmr build configurations. Still strange. I am not 100% convinced yet.

**Comment by mateuszviste** [comment]
The "invalid opcode" message suggests that *something* has overwritten some executable code in memory (possibly belonging to SvarCOM) with garbage. I'd be tempted to say this is a jwasm issue related to some pointer math that works in flat mode but fails in segmented mode and writes data at some unexpected location. Perhaps using SvarCOM and/or QEMU makes this bug more likely to happen due to their memory layout? (that might explain why a debug version works - it simply puts garbage in a different location that happens to be less vital)
In any case, it's the kind of issues that is extremely difficult to track down...

**Comment by boeckmann** [comment]
Yes, I tried to debug this. In this specific case, the invalid opcode occurs shortly after program start. I did not see what could go wrong up to this specific location. My guess is still that OpenWatcom generates some broken code in this specific case. It also seems to run under QEMU with slightly different build parameters. So I think I will replace the current JWASMR executable of the SvarDOS package with a build of mine that also runs under QEMU.
